<h1>Quiz 1 : Pemahaman ALgoritma</h1>

Jawab Pertanyaan di bawah ini dengan penyampaian bahasa sendiri!

1. Bagaimana Process algoritma KNN?
2. Jelaskan minimal 2 kekurangan dari Algoritma KNN?

1. kNN dilakukan dengan mencari kelompok k objek dalam data training yang paling dekat (mirip) dengan objek pada data baru atau data testing

2. - Perlu menunjukkan parameter K (jumlah tetangga terdekat) 
   - Tidak menangani nilai hilang (missing value) secara implisit 
   - Sensitif terhadap data pencilan (outlier) 
   - Rentan terhadap variabel yang non-informatif

<h1>Quiz 2 : Pengaplikasian Algoritma KNN</h1>

- Buatlah model KNN
- Training model tersebut kepada data X_train dengan hyperparaeter kombinasi antara  2<= k <=20 dan weights bernilai 'uniform', dan 'distance'  
- Lakukan prediksi terhadap data X_test untuk model di setiap kombinasi k dan weights
- Score dengan accuracy dari hasil prediksi terhadap data X_test untuk setiap kombinasi hyperparameter, masukan score tersebut ke dalam objek list dengan di groupkan berdasarkan jenis weights, sehingga di akhir kita punya variable misalkan list_uniform_score, dan list_distance_score
- Plot ke dua list tersebut dengan lineplot kedalam satu axis dan bandingkan hasilnya

In [1]:
from sklearn.datasets import load_wine
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

wine = load_wine()
df = pd.DataFrame(np.c_[wine['data'], wine['target']], columns= wine['feature_names'] + ['class'])

#Preprocessing data
X = df.drop('class', axis=1)
y = df['class']

from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from scipy import stats
from sklearn.metrics import explained_variance_score

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=21)

In [5]:
x = stats.zscore(X_train)

model_knn1 = KNeighborsClassifier(n_neighbors=2, weights='uniform')
model_knn2 = KNeighborsClassifier(n_neighbors=2, weights='distance')

model_knn1.fit(X_train, y_train)
y_tests = model_knn1.predict(X_test)
score = model_knn1.score(X_test, y_test)

model_knn2.fit(X_train, y_train)
y_test1 = model_knn2.predict(X_test)
score1 = model_knn2.score(X_test, y_test)

print('Uniform -> ', score)
print('Distance -> ', score1)

# fig, ax = plt.subplots(figsize=(16, 8))
# ax.plot(X, y)

# plt.show()

Uniform ->  0.6944444444444444
Distance ->  0.75


In [8]:
# Membuat model KNN
model_u = []
model_d = []
# menggunakan Looping for
for k_value in range(2, 21):
    model_u.append(KNeighborsClassifier(n_neighbors=k_value, weights='uniform'))
    model_d.append(KNeighborsClassifier(n_neighbors=k_value, weights='distance'))
                   
# model KNN belajar dari data (training)
for k_value in range(len(model_u)):
    model_u[k_value].fit(X_train, y_train[:len(X_train)])
    model_d[k_value].fit(X_train, y_train[:len(X_train)])
                   
#prediksi hasil test X
y_u = []
y_d = []
for k_value in range(len(model_u)):
    y_u.append(model_u[k_value].predict(X_test).mean())
    y_d.append(model_u[k_value].predict(X_test).mean())
                   
print('Uniform Score -> ', list(y_u), '\n')
print('Distance Score ->', list(y_d), '\n')

Uniform Score ->  [0.8055555555555556, 0.9444444444444444, 0.8055555555555556, 0.9444444444444444, 1.0, 0.9166666666666666, 0.8888888888888888, 1.0, 1.0, 1.0277777777777777, 0.9444444444444444, 0.9722222222222222, 0.9444444444444444, 0.9722222222222222, 0.9166666666666666, 0.9722222222222222, 0.9444444444444444, 0.9722222222222222, 0.9166666666666666] 

Distance Score -> [0.8055555555555556, 0.9444444444444444, 0.8055555555555556, 0.9444444444444444, 1.0, 0.9166666666666666, 0.8888888888888888, 1.0, 1.0, 1.0277777777777777, 0.9444444444444444, 0.9722222222222222, 0.9444444444444444, 0.9722222222222222, 0.9166666666666666, 0.9722222222222222, 0.9444444444444444, 0.9722222222222222, 0.9166666666666666] 



Contoh hasil yang di harapkan :

Uniform Scores -> [0.6944444444444444,
 0.8055555555555556,
 0.7222222222222222,
 0.6666666666666666,
 0.6388888888888888,
 0.75,
 0.6388888888888888,
 0.6944444444444444,
 0.6944444444444444,
 0.7222222222222222,
 0.6388888888888888,
 0.6666666666666666,
 0.6388888888888888,
 0.6666666666666666,
 0.6111111111111112,
 0.6666666666666666,
 0.6388888888888888,
 0.6666666666666666,
 0.6111111111111112]
 
 distance scores -> [0.75,
 0.8333333333333334,
 0.7777777777777778,
 0.75,
 0.7222222222222222,
 0.75,
 0.6944444444444444,
 0.7222222222222222,
 0.6944444444444444,
 0.6944444444444444,
 0.6944444444444444,
 0.6944444444444444,
 0.6944444444444444,
 0.6666666666666666,
 0.6666666666666666,
 0.6666666666666666,
 0.6666666666666666,
 0.6666666666666666,
 0.6666666666666666]

![image.png](attachment:image.png)

<h1>Quiz 3 : Pengaplikasian Model Linear Regression</h1>

Applikasikan model linear regresi kedalam data berikut. Kemudian tunjukan parameter koefesien dan intercept dari model tersebut.

In [ ]:
import numpy as np

rng = np.random.RandomState(1)

X = 10 * rng.rand(50, 4)
y = np.array([  0.9826564 ,  49.40390035,  85.76013175,  29.17254633,
        11.270054  ,  61.53852735, -19.52503854,  10.10230867,
        82.03058206,  45.72660678,  82.09252575,  90.78872391,
        67.94178098,  39.97492762, 124.78866966,  85.24615819,
        85.68491086,  80.39481211,  14.54591581, 137.19722354,
        85.04063428,  94.76681927,  84.64289989,  38.76421156,
        47.33953927,  94.50853335,  77.3276567 ,  -9.77846805,
        61.13434468,  42.12124052,  26.65634335, 120.41593333,
       138.08132504, 103.33745675, 145.64447692,  19.54777986,
        87.40631024, 111.47327389,  95.94874761, 146.34817502,
       118.10333528,  56.85956484, 108.57660235, 114.30790247,
        40.27107041,  73.83871008,  45.52217182,  -7.57752547,
       147.81772162, 139.40285349])